### PR TITLE
[Cute,Fwd] Extend score_mod to variable sequence length

### PR DESCRIPTION
**Pull request by reubenconducts**
This PR extends the flexattention-style `score_mod` support from #1840 to the variable sequence length setting. The following cases are now covered:
- When `q` and/or `k` are varlen in any combination (TF, FT, TT, FF)
- PackGQA for varlen on SM90 only 

This does incur a slight complexity overhead in defining `score_mod`s, but many cases -- particularly when not using auxiliary tensors -- no change is needed. We update the `score_mod` signature to pass in `seqlen_info`. This allows the user to access offsets within the `score_mod`, needed to index properly into packed auxiliary tensors. It also allows the user to access sequence lengths, making offset causal and other such mods simple. 

The expected signature for `score_mod`s is now `(scores, b_idx, h_idx, q_idx, kv_idx, seqlen_info, aux_tensors)`. 

CC: @jayhshah @drisspg

### CODE REVIEW

**Comment by tridao** [comment]
Is there a use case where one needs both q idx and q idx global? 
Does it make more sense to always have q idx being the idx wrt the current sequence (ie no q idx global)? Same w k idx

File: tests/cute/score_mod_definitions.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+    dtype = token_bias.element_type
+    q_frag = cute.make_fragment(1, cutlass.Int32)
+    q_frag.store(q_idx_global)
+    bias_frag = cute.make_fragment(1, dtype)
+    bias_frag[0] = token_bias[q_frag[0]]
+    return tSrS_ssa + (bias_frag.load()).to(cutlass.Float32)
+
+
+@cute.jit
+def score_mod_global_rel_plus_kv_bias(
```
Comment (reubenconducts) [inline tests/cute/score_mod_definitions.py:154]: @tridao For example here, when using both relative positional bias and per-token kv bias, we need both `kv_idx` and `kv_idx_global`.

File: flash_attn/cute/flash_fwd.py [hunk 2]
```diff
@@ -2400,6 +2411,25 @@ def apply_score_mod(
...
+        recompute_fastdiv_mods_k = cutlass.const_expr(
+            aux_tensors is not None and seqlen.has_cu_seqlens_k
+        )
+
+        if cutlass.const_expr(fastdiv_mods is not None):
+            seqlen_q_divmod, seqlen_k_divmod = fastdiv_mods
+            fastdiv_mods = (
+                seqlen_q_divmod
+                if not recompute_fastdiv_mods_q
+                else FastDivmodDivisor(seqlen.seqlen_q),
```
Comment (drisspg) [inline flash_attn/cute/flash_fwd.py]: noob q, doesnt the divisor need to be build on the host?

File: flash_attn/cute/flash_fwd_sm100.py [hunk 3]
```diff
@@ -1864,6 +1865,7 @@ def softmax_step(
         cute.arch.mbarrier_wait(mbar_ptr + self.mbar_S_full_offset + stage, mma_si_consumer_phase)
         tSrS_t2r = cute.make_fragment(thr_tmem_load.partition_D(tScS).shape, self.qk_acc_dtype)
         cute.copy(thr_tmem_load, tStS_t2r, tSrS_t2r)
+        # cute.arch.sync_warp()
```
Comment (drisspg) [inline flash_attn/cute/flash_fwd_sm100.py]: *

File: flash_attn/cute/flash_fwd_sm100.py [hunk 4]
```diff
@@ -2637,30 +2636,47 @@ def apply_score_mod(
...
+            # Extract the logical q position and head offset
+            q_idx_logical_adjusted = q_idx_logical // self.qhead_per_kvhead
+            head_offset = q_idx_logical - q_idx_logical_adjusted * self.qhead_per_kvhead
             head_idx = head_idx * self.qhead_per_kvhead + head_offset
+            q_idx_logical = q_idx_logical_adjusted
 
-        if cutlass.const_expr(aux_tensors is not None):
-            seqlen_q_divmod, _ = fastdiv_mods
-            _, q_idx_logical = divmod(q_idx_logical, seqlen_q_divmod)
+        # Recompute fastdiv_mods if necessary
```
Comment (drisspg) [inline flash_attn/cute/flash_fwd_sm100.py]: whats going on here?

File: flash_attn/cute/interface.py [hunk 5]
```diff
@@ -346,7 +341,32 @@ def _flash_attn_fwd(
     elif lse is not None:
         lse_tensor = from_dlpack(lse.detach(), assumed_align=4).mark_layout_dynamic(leading_dim=lse.ndim - 1)
     else:
-        lse_tensor = None 
+        lse_tensor = None
+
+    def maybe_wrap_score_mod(score_mod):
```
Comment (drisspg) [inline flash_attn/cute/interface.py]: nit: can we pull into utils?

File: flash_attn/cute/interface.py [hunk 6]
```diff
@@ -346,7 +341,32 @@ def _flash_attn_fwd(
...
+        """wrap score_mod to accept global indices, needed for varlen sequences"""
+        if score_mod is None:
+            return None
+
+        sig = inspect.signature(score_mod)
```
Comment (drisspg) [inline flash_attn/cute/interface.py]: I think this might be host side slow, I have an issue to update inspect usage in utils.hash_callable

File: flash_attn/cute/softmax.py [hunk 7]
```diff
@@ -29,8 +30,8 @@ def create(
         arch: cutlass.Constexpr[int] = 80,
         softmax_scale: Float32 | None = None,
     ):
-        row_max = cute.make_fragment(num_rows, Float32)
-        row_sum = cute.make_fragment(num_rows, Float32)
+        row_max = cute.make_rmem_tensor(num_rows, Float32)
```
Comment (drisspg) [inline flash_attn/cute/softmax.py:33]: thank you been meaning to do this

**Comment by reubenconducts** [comment]
@drisspg major refactor addressing many points we discussed

File: flash_attn/cute/flash_fwd.py (hunk 2 above)
Comment (reubenconducts) [reply flash_attn/cute/flash_fwd.py]: no, it's okay on device

File: flash_attn/cute/flash_fwd_sm100.py (hunk 4 above)
Comment (reubenconducts) [reply flash_attn/cute/flash_fwd_sm100.py]: simplified; now `q_idx_logical` computation is after possible divmod recomputation

File: flash_attn/cute/interface.py (hunk 6 above)
Comment (reubenconducts) [reply flash_attn/cute/interface.py]: removed
